File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void DFS(int u,vector<int>&vis,unordered_map<int,vector<int>>&adj){
        vis[u]=true;
        for(int &v:adj[u]){
            if(!vis[v]) DFS(v,vis,adj);
        }
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        unordered_map<int,vector<int>>adj;
        for(int i=0;i<isConnected.size();i++){
            for(int j=0;j<isConnected[0].size();j++){
                if(isConnected[i][j]==1){
                    adj[i].push_back(j);
                }
            }
        }

        vector<int>vis(isConnected.size(),0);
        int count=0;
        for(int i=0;i<isConnected.size();i++){
            if(!vis[i]){
                DFS(i,vis,adj);
                count++;
            }
        }
        return count;
    }
};

File: 547-number-of-provinces/number-of-provinces.py
class Solution:
    def findParent(self,x):
        if(self.parent[x]==x):
            return x
        self.parent[x]=self.findParent(self.parent[x])
        return self.parent[x]

    def Union(self,x,y):
        x = self.findParent(x)
        y = self.findParent(y)

        if x == y: 
            return

        if(self.rank[x]<self.rank[y]):
            self.parent[x]=y
        elif (self.rank[x]>self.rank[y]):
            self.parent[y]=x
        else:
            self.parent[y]=x
            self.rank[x]+=1

        self.components-=1       

    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        self.components = n
        self.parent=[i for i in range(n)]
        self.rank = [0]*n

        for i in range(n):
            for j in range(n):
                if isConnected[i][j]==1:
                    self.Union(i,j)

        return self.components



        